pkg/gatewayconfigurationserver/gatewayconfigurationserver.go

```diff
@@ -0,0 +1,155 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gatewayconfigurationserver
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/gogo/protobuf/types"
+	echo "github.com/labstack/echo/v4"
+	"go.thethings.network/lorawan-stack/pkg/basicstation/cups"
+	"go.thethings.network/lorawan-stack/pkg/component"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/pfconfig/semtechudp"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	ttnweb "go.thethings.network/lorawan-stack/pkg/web"
+	"google.golang.org/grpc/metadata"
+)
+
+// Config contains the Gateway Configuration Server configuration.
+type Config struct {
+	// BasicStation defines the configuration for the BasicStation CUPS server.
+	BasicStation cups.ServerConfig `name:"basic-station" description:"BasicStation CUPS server configuration."`
+	// RequreAuth defines if the HTTP endpoints should require authentication or not.
+	RequireAuth bool `name:"require-auth" description:"Require authentication for the HTTP endpoints."`
+}
+
+// GatewayConfigurationServer implements the Gateway Configuration Server component.
+type GatewayConfigurationServer struct {
+	*component.Component
+	config *Config
+
+	registry ttnpb.GatewayRegistryClient
+
+	ctx context.Context
+}
+
+// RegisterRoutes registers the web frontend routes.
+func (gcs *GatewayConfigurationServer) RegisterRoutes(server *ttnweb.Server) {
+	middleware := []echo.MiddlewareFunc{
+		gcs.handleError(),
+		gcs.validateAndFillIDs(),
+	}
+	if gcs.config.RequireAuth {
+		middleware = append(middleware, gcs.requireGatewayRights(ttnpb.RIGHT_GATEWAY_INFO))
+	}
+	group := server.Group(ttnpb.HTTPAPIPrefix+"/gcs/gateways/:gateway_id", middleware...)
+	group.GET("/global_conf.json", func(c echo.Context) error {
+		return gcs.handleGet(c)
+	})
+}
+
+// New returns new *GatewayConfigurationServer.
+func New(c *component.Component, conf *Config, opts ...Option) (*GatewayConfigurationServer, error) {
+	gcs := &GatewayConfigurationServer{
+		Component: c,
+		config:    conf,
+		ctx:       c.FillContext(c.Context()),
```

Comment:
Is this necessary? I thought the component's context already had everything filled that needed to be filled. As far as I know the `FillContext` was meant for filling request contexts that didn't pass the interceptor/echo middleware. cc: @johanstokking